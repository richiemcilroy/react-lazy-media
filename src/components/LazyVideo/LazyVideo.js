import React, {Component} from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import {ReactLazyMedia} from '../../ReactLazyMedia';

export class LazyVideo extends Component {
  static propTypes = {
    classes: PropTypes.arrayOf(PropTypes.string),
    controls: PropTypes.bool,
    id: PropTypes.string,
    src: PropTypes.string.isRequired,
    width: PropTypes.number,
    height: PropTypes.number,
    poster: PropTypes.string,
    autoplay: PropTypes.bool,
    preload: PropTypes.bool,
    muted: PropTypes.bool,
    loop: PropTypes.bool
  };

  static defaultProps = {
    classes: [],
    controls: true,
    autoplay: false,
    preload: false,
    loop: false,
    muted: true
  };

  // Update lazyLoad after first rendering of every image
  componentDidMount() {
    ReactLazyMedia.getInstance();
  }

  render() {
    const {
      classes,
      id,
      controls,
      width,
      height,
      src,
      poster,
      autoplay,
      preload,
      muted,
      loop,
      forwardedRef
    } = this.props;
    const mediaType = src.split('.').pop();

    //Only supporting mp4 and webm (for now?)
    if (mediaType !== 'mp4' && mediaType !== 'webm') {
      return null;
    }

    //TODO: Add support for autoplaying videos

    return (
      <video
        ref={forwardedRef}
        className={[classNames('lazy', ...classes)]}
        id={id}
        width={width}
        height={height}
        data-poster={poster}
        controls={controls}
        autoPlay={autoplay}
        muted={muted}
        loop={loop}
        preload={preload ? 'auto' : 'none'}
      >
        <source src={src} type={`video/${mediaType}`} />
      </video>
    );
  }
}

export const LazyVideoWithRef = React.forwardRef((props, ref) => <LazyVideo forwardedRef={ref} {...props} />);
