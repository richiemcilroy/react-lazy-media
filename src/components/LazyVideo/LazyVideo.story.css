.lazyVideo {
  margin-top: 150vh;
}
