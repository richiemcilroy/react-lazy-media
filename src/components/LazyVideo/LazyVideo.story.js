import React from 'react';
import {storiesOf} from '@storybook/react';
import {withKnobs, text} from '@storybook/addon-knobs';
import {LazyVideo} from './LazyVideo';
import './LazyVideo.story.css';

storiesOf('Lazy', module)
  .addDecorator(withKnobs)
  .add('LazyVideo mp4 without poster', () => {
    return (
      <LazyVideo
        classes={['lazyVideo']}
        src={
          'https://d2v9y0dukr6mq2.cloudfront.net/video/preview/Btw9Lp04ipguag6m/videoblocks-puppies-of-a-golden-retriever-are-playing-in-the-park_h2uzyqqaz__SB_PM.mp4'
        }
      />
    );
  })
  .add('LazyVideo mp4 with poster', () => {
    return (
      <LazyVideo
        classes={['lazyVideo']}
        src={
          'https://d2v9y0dukr6mq2.cloudfront.net/video/preview/Btw9Lp04ipguag6m/videoblocks-puppies-of-a-golden-retriever-are-playing-in-the-park_h2uzyqqaz__SB_PM.mp4'
        }
        poster={'http://placekitten.com/200/300'}
        preload={false}
        height={300}
        width={300}
      />
    );
  })
  .add('LazyVideo webm with poster', () => {
    return (
      <LazyVideo
        classes={['lazyVideo']}
        src={'https://d1ow200m9i3wyh.cloudfront.net/videos/MP_BG_Video_06.webm'}
        poster={'http://placekitten.com/200/300'}
        preload={false}
      />
    );
  });
