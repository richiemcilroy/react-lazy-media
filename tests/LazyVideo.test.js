import React from 'react';
import renderer from 'react-test-renderer';
import {LazyVideo} from '../src/components/LazyVideo/LazyVideo';

test('LazyVideo should require a src prop', () => {
  expect(() => {
    renderer.create(<LazyVideo />);
  }).toThrowErrorMatchingSnapshot();
});

test('LazyVideo should have default props', () => {
  const component = renderer.create(<LazyVideo src={'/video.mp4'} />);

  const tree = component.toJSON();

  expect(tree.props).toEqual({
    'data-src': '/video.mp4',
    autoPlay: false,
    controls: true,
    loop: false,
    muted: true,
    preload: 'none',
    className: 'lazy'
  });

  const child = tree.children.pop();

  expect(child.type).toBe('source');

  expect(child.props).toEqual({
    src: '/video.mp4',
    type: 'video/mp4'
  });
});

test('LazyVideo should render a video with a src', () => {
  const component = renderer.create(<LazyVideo src={'/video.mp4'} />);
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('LazyVideo should render a video with a src and poster', () => {
  const component = renderer.create(<LazyVideo src={'/video.mp4'} poster={'/poster.jpg'} />);
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('LazyVideo should be preloaded', () => {
  const component = renderer.create(<LazyVideo src={'/video.mp4'} preload={true} />);
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('LazyVideo should not be preloaded', () => {
  const component = renderer.create(<LazyVideo src={'/video.mp4'} preload={false} />);
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('LazyVideo should not be muted', () => {
  const component = renderer.create(<LazyVideo videoUrl={'/video.mp4'} mute={false} />);
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('LazyVideo should not show controls', () => {
  const component = renderer.create(<LazyVideo src={'/video.mp4'} showControls={false} />);
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('LazyVideo should autoplay', () => {
  const component = renderer.create(<LazyVideo src={'/video.mp4'} autoPlay={true} />);
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('LazyVideo should have 300px width', () => {
  const component = renderer.create(<LazyVideo src={'/video.mp4'} width={300} />);
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('LazyVideo should have 300px height', () => {
  const component = renderer.create(<LazyVideo src={'/video.mp4'} height={300} />);
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('LazyVideo should have an id', () => {
  const component = renderer.create(<LazyVideo src={'/video.mp4'} id={'lazyVideo'} />);
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('LazyVideo should have classes', () => {
  const component = renderer.create(<LazyVideo src={'/video.mp4'} classes={['class1', 'class2']} />);
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
