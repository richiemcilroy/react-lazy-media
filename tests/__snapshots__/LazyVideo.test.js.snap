// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LazyVideo should autoplay 1`] = `
<video
  autoPlay={false}
  className={
    Array [
      "lazy",
    ]
  }
  controls={true}
  loop={false}
  muted={true}
  preload="none"
>
  <source
    src="/video.mp4"
    type="video/mp4"
  />
</video>
`;

exports[`LazyVideo should be preloaded 1`] = `
<video
  autoPlay={false}
  className={
    Array [
      "lazy",
    ]
  }
  controls={true}
  loop={false}
  muted={true}
  preload="auto"
>
  <source
    src="/video.mp4"
    type="video/mp4"
  />
</video>
`;

exports[`LazyVideo should have 300px height 1`] = `
<video
  autoPlay={false}
  className={
    Array [
      "lazy",
    ]
  }
  controls={true}
  height={300}
  loop={false}
  muted={true}
  preload="none"
>
  <source
    src="/video.mp4"
    type="video/mp4"
  />
</video>
`;

exports[`LazyVideo should have 300px width 1`] = `
<video
  autoPlay={false}
  className={
    Array [
      "lazy",
    ]
  }
  controls={true}
  loop={false}
  muted={true}
  preload="none"
  width={300}
>
  <source
    src="/video.mp4"
    type="video/mp4"
  />
</video>
`;

exports[`LazyVideo should have an id 1`] = `
<video
  autoPlay={false}
  className={
    Array [
      "lazy",
    ]
  }
  controls={true}
  id="lazyVideo"
  loop={false}
  muted={true}
  preload="none"
>
  <source
    src="/video.mp4"
    type="video/mp4"
  />
</video>
`;

exports[`LazyVideo should have classes 1`] = `
<video
  autoPlay={false}
  className={
    Array [
      "lazy class1 class2",
    ]
  }
  controls={true}
  loop={false}
  muted={true}
  preload="none"
>
  <source
    src="/video.mp4"
    type="video/mp4"
  />
</video>
`;

exports[`LazyVideo should not be muted 1`] = `
<video
  autoPlay={false}
  className={
    Array [
      "lazy",
    ]
  }
  controls={true}
  loop={false}
  muted={true}
  preload="none"
>
  <source
    src="/video.mp4"
    type="video/mp4"
  />
</video>
`;

exports[`LazyVideo should not be preloaded 1`] = `
<video
  autoPlay={false}
  className={
    Array [
      "lazy",
    ]
  }
  controls={true}
  loop={false}
  muted={true}
  preload="none"
>
  <source
    src="/video.mp4"
    type="video/mp4"
  />
</video>
`;

exports[`LazyVideo should not show controls 1`] = `
<video
  autoPlay={false}
  className={
    Array [
      "lazy",
    ]
  }
  controls={true}
  loop={false}
  muted={true}
  preload="none"
>
  <source
    src="/video.mp4"
    type="video/mp4"
  />
</video>
`;

exports[`LazyVideo should render a video with a src 1`] = `
<video
  autoPlay={false}
  className={
    Array [
      "lazy",
    ]
  }
  controls={true}
  loop={false}
  muted={true}
  preload="none"
>
  <source
    src="/video.mp4"
    type="video/mp4"
  />
</video>
`;

exports[`LazyVideo should render a video with a src and poster 1`] = `
<video
  autoPlay={false}
  className={
    Array [
      "lazy",
    ]
  }
  controls={true}
  data-poster="/poster.jpg"
  loop={false}
  muted={true}
  preload="none"
>
  <source
    src="/video.mp4"
    type="video/mp4"
  />
</video>
`;

exports[`LazyVideo should require a src prop 1`] = `
"Warning: Failed prop type: The prop \`src\` is marked as required in \`LazyVideo\`, but its value is \`undefined\`.
    in LazyVideo"
`;
